sdks/java/core/src/main/java/org/apache/beam/sdk/testing/TestPipelineRule.java

```diff
@@ -0,0 +1,121 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.beam.sdk.testing;
+
+import org.apache.beam.sdk.Pipeline;
+import org.apache.beam.sdk.PipelineResult;
+import org.apache.beam.sdk.options.PipelineOptions;
+import org.apache.beam.sdk.runners.PipelineRunner;
+import org.apache.beam.sdk.runners.TransformHierarchy;
+import org.junit.rules.TestRule;
+import org.junit.runner.Description;
+import org.junit.runners.model.Statement;
+
+/**
+ * A {@link TestRule} that verifies the following assumptions are met upon running a test that
+ * involves a {@link TestPipeline}:
+ * <ul>
+ *   <li>PAssert is not added to a pipeline that has already been run.
+ *   <li>The pipeline is run after the PAssert is added.
+ * </ul>
+ */
+public class TestPipelineRule extends TestPipeline implements TestRule {
```

Comment:
You're right in that we shouldn't always have a hard failure if there is no `PAssert` node - certainly some tests that run with `TestPipeline` could use a non-`PAssert` way of asserting correct behavior (e.g., with expected failure in [`PAssertTest`](https://github.com/apache/incubator-beam/blob/master/sdks/java/core/src/test/java/org/apache/beam/sdk/testing/PAssertTest.java#L401)). Some tests will use `TestPipeline` only to construct a graph - we need to expose some mutator that we're just using as a construction pipeline
 
```
@Rule ExpectedException ex = ExpectedException.none();
...
ex.expect(FooException.class)
```
We can have
```
@Rule TestPipeline p = TestPipeline.create();
...
p.noRun()
```
(naming TBD)

The goal is ultimately to move all uses of `TestPipeline` to be via `TestRule` as "the right way".

I don't think it would be too bad to start with "when used as a rule will always enforce that the pipeline is run" and migrate to be more flexible in the future though.